Introduce the concept of Frontend Attributes.

AnthonyBarbier opened the pull request:
Summary:
    Frontend Attributes can be set by the user or the frontend and
    are passed through to the XLA backend as a dictionary of strings where they can
    be used to modify the way the HLO instructions are executed.

XLA Development discussion:
    https://groups.google.com/d/msg/xla-dev/9TM0-1N_JlM/Q2R8o2RgBwAJ

Test Plan:
    bazel test returned:
    INFO: Executed 522 out of 522 tests: 522 tests pass.
    INFO: There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.
    INFO: Build completed successfully, 4027 total actions
    SUCCESS!

AnthonyBarbier committed:
Introduce the concept of Frontend Attributes.

Summary:
    Frontend Attributes can be set by the user or the frontend and
    are passed through to the XLA backend as a dictionary of strings where they can
    be used to modify the way the HLO instructions are executed.

XLA Development discussion:
    https://groups.google.com/d/msg/xla-dev/9TM0-1N_JlM/Q2R8o2RgBwAJ

Test Plan:
    bazel test returned:
    INFO: Executed 522 out of 522 tests: 522 tests pass.
    INFO: There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.
    INFO: Build completed successfully, 4027 total actions
    SUCCESS!

sanjoy commented on tensorflow/compiler/tf2xla/frontend_attributes_util.cc:
```diff
@@ -0,0 +1,43 @@
+xla::StatusOr<absl::optional<xla::FrontendAttributes>>
+GetFrontendAttributesFromNodeDef(const NodeDef& node_def) {
+  if (!HasNodeAttr(node_def, kFrontendAttributesAttribute)) {
+    return absl::optional<xla::FrontendAttributes>();
```
Can you return `absl::nullopt`?

AnthonyBarbier replied:
I can't just do `return absl::nullopt;` because the `optional` is embedded inside a `StatusOr`:
```
ERROR: tensorflow/compiler/tf2xla/frontend_attributes_util.cc:29:12: error: could not convert 'absl::nullopt' from 'absl::internal::identity_t<absl::nullopt_t> {aka const absl::nullopt_t}' to 'xla::StatusOr<absl::optional<xla::FrontendAttributes> > {aka stream_executor::port::StatusOr<absl::optional<xla::FrontendAttributes> >}'
ERROR:      return absl::nullopt;
```
The alternative is to do: `return xla::StatusOr<absl::optional<xla::FrontendAttributes>>(absl::nullopt);` but I think it ends up being the same as `return absl::optional<xla::FrontendAttributes>();` 

https://github.com/abseil/abseil-cpp/blob/master/absl/types/optional.h#L130

I'll change it anyway if you think it's more obvious with the `nullopt`

sanjoy commented on tensorflow/compiler/tf2xla/frontend_attributes_util.cc:
```diff
@@ -0,0 +1,43 @@
+        "Experimental _XlaFrontendAttributes attribute was not a valid encoded "
+        "xla::FrontendAttributes proto.");
+  }
+  return absl::optional<xla::FrontendAttributes>(attributes);
```
Can this be just `return attributes;`?

AnthonyBarbier replied:
No I don't think it can, probably for the same reason as above: the optional is embedded inside a StatusOr and as a result the conversion needs to be explicit:
```
ERROR: tensorflow/compiler/tf2xla/frontend_attributes_util.cc: In function 'xla::StatusOr<absl::optional<xla::FrontendAttributes> > tensorflow::GetFrontendAttributesFromNodeDef(const tensorflow::NodeDef&)':
ERROR: tensorflow/compiler/tf2xla/frontend_attributes_util.cc:40:10: error: could not convert 'attributes' from 'xla::FrontendAttributes' to 'xla::StatusOr<absl::optional<xla::FrontendAttributes> > {aka stream_executor::port::StatusOr<absl::optional<xla::FrontendAttributes> >}'
ERROR:    return attributes;
```

sanjoy commented on tensorflow/compiler/xla/client/xla_builder.h:
```diff
@@ -1038,6 +1056,37 @@ class XlaScopedShardingAssignment {
+      const XlaScopedFrontendAttributesAssignment&) = delete;
+
+  ~XlaScopedFrontendAttributesAssignment() {
+    SetFrontendAttributes(absl::nullopt);
```
Shouldn't this restore the frontend attributes to the value before it was set (i.e. before `XlaScopedFrontendAttributesAssignment::XlaScopedFrontendAttributesAssignment`)?

AnthonyBarbier replied:
That's a good point, I guess we need to decide how these attributes are getting set.
Currently, we use them internally to set attributes at the python scope level that will affect the operations added in that scope, so we do something like:
```python
@tf_contextlib.contextmanager
def frontend_attribute(attribute_name, attribute_value):

  proto = xla_data_pb2.FrontendAttributes()
  # pylint: disable=protected-access
  attr_value = ops.get_default_graph()._attr_scope_map.get(
      FRONTEND_ATTRIBUTES_NAME)
  # pylint: enable=protected-access
  new_attribute = True
  if attr_value:
    proto.ParseFromString(attr_value.s)
  proto.map[attribute_name] = attribute_value

  attr_value = attr_value_pb2.AttrValue(s=proto.SerializeToString())
  attrs = {FRONTEND_ATTRIBUTES_NAME: attr_value}

  # pylint: disable=protected-access
  with ops.get_default_graph()._attr_scope(attrs):
    yield
  # pylint: enable=protected-access
```
So basically the backup / restore of the previous attributes is done at the python level and in `XlaCompilationDevice::Compute` we could just do `builder->SetFrontendAttributes()` the ScopeAssignment is actually overkill.
Our C++ backend also adds some attributes, but this all happens further down the pipeline so it's not a problem here.

Now, I guess at some point we might want to set `FrontendAttributes` from the C++ runtime before `XlaCompilationDevice::Compute`.
If we consider this a valid use case then we would indeed need a ScopeAssignment but I think in that case it should probably do:

ScopeAssignment constructor:
- save existing cpp attributes.
- merge the python attributes with the cpp attributes and set them as the current attributes in the builder.

ScopeAssignment destructor:
- restore cpp attributes.

Otherwise with the current set/clear calls we will randomly erase the cpp attributes whenever some are set in python.

sanjoy replied:
> If we consider this a valid use case then we would indeed need a ScopeAssignment but I think in that case it should probably do:
> 
> ScopeAssignment constructor:
> 
> * save existing cpp attributes.
> * merge the python attributes with the cpp attributes and set them as the current attributes in the builder.

I don't think the merging behavior should be implemented by `XlaScopedFrontendAttributesAssignment`.  It should be implemented in its caller.

And `XlaScopedFrontendAttributesAssignment` should just save/restore frontend attributes without any intelligent merging behavior.

IOW the caller would do:

```
FrontendAttrs fa = GetCurrentAttrs();
FrontendAttrs new_fa = Merge(fa, attrs_from_node_def);
XlaScopedFrontendAttributesAssignment raii(std::move(new_fa));
```

Naively this would incur an extra copy, but that can be avoided by making `XlaScopedFrontendAttributesAssignment` move out the existing `FrontendAttrs` instead of copying it out.

> 
> ScopeAssignment destructor:
> 
> * restore cpp attributes.
> 
> Otherwise with the current set/clear calls we will randomly erase the cpp attributes whenever some are set in python.

AnthonyBarbier replied:
So basically you want to replace `XlaBuilder::SetFrontendAttributes` with a `XlaBuilder::SwapFrontendAttributes` that will return the old value in order to avoid the copy ?

sanjoy replied:
Yeah, that would work as well (and sounds simpler than my idea of adding a `FrontendAttributes&& ConsumeFrontendAttributes()`).

AnthonyBarbier replied:
Done

sanjoy commented on tensorflow/compiler/xla/client/xla_builder.h:
```diff
@@ -158,6 +158,16 @@ class XlaBuilder {
   // Sets an OpSharding that will be attached to all instructions until cleared.
   void SetSharding(const OpSharding& sharding) { sharding_ = sharding; }
 
+  void SetFrontendAttributes(const FrontendAttributes& frontend_attributes) {
```
Add comments.

AnthonyBarbier replied:
Done

sanjoy commented on tensorflow/compiler/xla/client/xla_builder.cc:
```diff
@@ -289,6 +289,14 @@ Status XlaBuilder::SetDynamicBinding(int64 dynamic_size_param_num,
   return Status::OK();
 }
 
+Status XlaBuilder::AddFrontendAttribute(const XlaOp& op, std::string attribute,
```
The declaration needs a comment that this overwrites existing attributes.

AnthonyBarbier replied:
Done

sanjoy commented on tensorflow/compiler/xla/client/xla_builder.cc:
```diff
@@ -289,6 +289,14 @@ Status XlaBuilder::SetDynamicBinding(int64 dynamic_size_param_num,
+                                        std::string value) {
+  TF_ASSIGN_OR_RETURN(auto instr_proto, LookUpMutableInstruction(op));
+  auto* frontend_attributes = instr_proto->mutable_frontend_attributes();
+  (*frontend_attributes->mutable_map())[attribute] = value;
```
Use `std::move`.

AnthonyBarbier replied:
Done

sanjoy commented on tensorflow/compiler/xla/client/xla_builder.h:
```diff
@@ -314,6 +324,10 @@ class XlaBuilder {
 
+  // Looks up the HloInstruction and sets the frontend attribute "attribute" to
+  // "value".
+  Status AddFrontendAttribute(const XlaOp& op, string attribute, string value);
```
Why do we need both this and `SetFrontendAttributes` etc.?

AnthonyBarbier replied:
Updating the doc.

Also I was thinking about renaming it `SetInstructionFrontendAttribute()` as it only sets the attribute attached to the instruction and not the builder's one.

sanjoy commented on tensorflow/compiler/xla/service/hlo_instruction.h:
```diff
@@ -1384,6 +1384,14 @@ class HloInstruction {
   }
   Status set_backend_config(const tensorflow::protobuf::Message& proto);
 
+  void set_frontend_attributes(FrontendAttributes frontend_attributes) {
```
We'll also need to serialize and parse these.  Are you going to add support for parsing / serializing in subsequent commits?

AnthonyBarbier replied:
Sorry, I'm not familiar with the Serialization API: I'm going to add the frontend attributes to the `ToProto` method, anything else there is to do ?

sanjoy replied:
See hlo_parser.h and hlo_parser.cc and `HloModule::ToString`

AnthonyBarbier replied:
Done

sanjoy commented on tensorflow/compiler/xla/service/hlo_instruction.h:
```diff
@@ -1878,6 +1886,10 @@ class HloInstruction {
 
+  // Attributes passed from the frontend to give hints to the backend about
+  // how to compile this HLO.
+  FrontendAttributes frontend_attributes_;
```
We need to state that HLO->HLO transforms are expected to preserve these attributes on a "best effort" basis only.  An example would also help.

AnthonyBarbier replied:
Done

sanjoy commented on tensorflow/compiler/tf2xla/frontend_attributes_util.h:
```diff
@@ -0,0 +1,32 @@
+namespace tensorflow {
+
+xla::StatusOr<absl::optional<xla::FrontendAttributes>>
+GetFrontendAttributesFromNodeDef(const NodeDef& node_def);
```
It is more idiomatic to take an `AttrSlice`.

AnthonyBarbier replied:
Done

AnthonyBarbier committed:
Addressed sanjoy's comments

AnthonyBarbier committed:
Renamed GetFrontendAttributesFromNodeDef -> GetFrontendAttributesFromAttrSlice and added documentation

sanjoy commented on tensorflow/compiler/tf2xla/frontend_attributes_util.cc:
```diff
@@ -24,15 +24,14 @@ const char kFrontendAttributesAttribute[] = "_XlaFrontendAttributes";
-  if (!HasNodeAttr(node_def, kFrontendAttributesAttribute)) {
-    return absl::optional<xla::FrontendAttributes>();
+GetFrontendAttributesFromNodeDef(const AttrSlice& attrs) {
+  auto attr = attrs.Find(kFrontendAttributesAttribute);
```
Preferable to use concrete type here, the type isn't obvious from the RHS.

AnthonyBarbier replied:
Done

sanjoy commented on tensorflow/compiler/xla/client/xla_builder.h:
```diff
@@ -158,14 +158,31 @@ class XlaBuilder {
+  // Sets the FrontendAttributes that will be added to all instructions until
+  // cleared.
+  //
+  // FrontendAttributes are often applied to a serie of XLA HLO instructions.
```
series

AnthonyBarbier replied:
Done

sanjoy commented on tensorflow/compiler/xla/client/xla_builder.h:
```diff
@@ -158,14 +158,31 @@ class XlaBuilder {
+  // cleared.
+  //
+  // FrontendAttributes are often applied to a serie of XLA HLO instructions.
+  // As a result they are set on the Computation Builder and all the
```
computation builder

AnthonyBarbier replied:
Done

sanjoy commented on tensorflow/compiler/xla/client/xla_builder.cc:
```diff
@@ -2719,32 +2729,69 @@ void XlaBuilder::AddCalledComputation(const XlaComputation& computation,
+template <typename InstructionType, typename HandleToIndexType,
+          typename InstructionProtoVectorType>
+StatusOr<InstructionType> LookUpInstructionByHandleInternal(
+    HandleToIndexType& handle_to_index,
```
Should this be `const HandleToIndexType &`

sanjoy replied:
Same for `InstructionProtoVectorType`.

AnthonyBarbier replied:
Done (requires a `const_cast` though for `InstructionProtoVectorType`)

AnthonyBarbier committed:
Rework #2

- Remove auto
- Remove templates / keep parameters const
- Fix comments

AnthonyBarbier committed:
Fixed case: Computation Builder -> computation builder

AnthonyBarbier committed:
Moved the attributes merging out of the ScopeAssignment class and changed MergeFrontendAttributes into SwapFrontendAttributes

AnthonyBarbier committed:
Added serializers / parsers and test

AnthonyBarbier committed:
Apply linter

sanjoy commented on tensorflow/compiler/xla/service/hlo_instruction.cc:
```diff
@@ -2538,6 +2550,9 @@ HloInstructionProto HloInstruction::ToProto() const {
       proto.mutable_outer_dimension_partitions()->Add(idx);
     }
   }
+  if (!frontend_attributes_.map().empty()) {
```
Why do we need to check `!frontend_attributes_.map().empty()`?

AnthonyBarbier replied:
Done

sanjoy reviewed (changes requested):
Looks fine to me modulo the nits.

CC @hyouklee in case he has some comments.

hyouklee commented:
Maybe I'm missing the context, but was there a reason why we couldn't reuse backend_config?

AnthonyBarbier commented:
You've got more context in the XLA-dev thread here : https://groups.google.com/d/msg/xla-dev/9TM0-1N_JlM/Q2R8o2RgBwAJ

Basically we do store the attributes as part of our backend config once we're on the XLA backend side but before that the backend config seems to already be used to store other information.

hyouklee commented:
What other information was stored in backend_config before the backend compiler starts? I only know the situation where the XLA client sets backend_config for the custom call string. Also, don't you have the same problem when moving some of the front-end config to the back-end config if it was already being used?

I'm fine with approving this change, but the frontend_attribute added in this PR seemed to mean almost identical to backend_config, with the same level of guarantee (i.e., we do not guarantee to preserve this during compiler transformations), and also the email discussion was about reusing the  backend_config.

AnthonyBarbier commented:
From a quick look through the code base it's being used by the `gpu` and the `cudnn` backends and also by the `custom_calls` since the `opaque` attribute was removed.
Seems risky to re-purpose an attribute which is already used in several places for different things.

AnthonyBarbier commented on tensorflow/compiler/xla/client/xla_builder.h:
```diff
@@ -1038,6 +1080,30 @@ class XlaScopedShardingAssignment {
+      saved_ = builder_->SwapFrontendAttributes(std::move(attributes));
+  }
+
+  XlaScopedFrontendAttributesAssignment(
```
Done

AnthonyBarbier commented on tensorflow/compiler/xla/client/xla_builder_test.cc:
```diff
@@ -978,5 +978,150 @@ TEST_F(XlaBuilderTest, CheckInputOutputAlias) {
   EXPECT_EQ(*alias_p1, ShapeIndex({0}));
 }
 
+void CheckAttributesMatch(const FrontendAttributes& attr,
```
Done

AnthonyBarbier commented on tensorflow/compiler/xla/client/xla_builder_test.cc:
```diff
@@ -978,5 +978,150 @@ TEST_F(XlaBuilderTest, CheckInputOutputAlias) {
+  }
+}
+
+void CheckInstructionsAttributesMatch(
```
Done

AnthonyBarbier committed:
Renamed Check -> Expect, used macro to disable copy / assignment

sanjoy commented:
Moreover, the backend config is not an uninterpreted map from strings to strings; it is a serialized proto where the type of the proto is backend dependent.

sanjoy commented:
@hyouklee WDYT?  Do you find the comments made here (for not using the already existing `backend_config`) convincing?

hyouklee commented:
Oh, I thought this PR was already merged in and closed. Yes, I'm fine with adding the front-end attribute.

sanjoy commented on tensorflow/compiler/xla/service/hlo_instruction.cc:
```diff
@@ -3190,6 +3204,12 @@ StatusOr<HloInstruction::FusionKind> StringToFusionKind(
 
+string FrontendAttributesToString(
+    const FrontendAttributes& frontend_attributes) {
+  return absl::StrFormat("{%s}", absl::StrJoin(frontend_attributes.map(), ",",
```
I don't think we can rely on a stable iteration order for `frontend_attributes.map()`, so this will give us non-deterministic serialization of HLO instructions.

AnthonyBarbier replied:
Done

sanjoy reviewed (changes requested):
LGTM with one minor nit inline.

AnthonyBarbier committed:
Sort map before serializing it

AnthonyBarbier committed:
- Fix: Save frontend attributes in while loop
- Fix: save backend / frontend attributes in ReplaceInstruction

sanjoy commented on tensorflow/compiler/xla/service/hlo_instruction.cc:
```diff
@@ -3190,6 +3204,15 @@ StatusOr<HloInstruction::FusionKind> StringToFusionKind(
+    const FrontendAttributes& frontend_attributes) {
+  std::vector<std::pair<string, string>> sorted_attributes(
+      frontend_attributes.map().begin(), frontend_attributes.map().end());
+  std::sort(sorted_attributes.begin(), sorted_attributes.end());
```
`absl::c_sort`?

AnthonyBarbier committed:
Use absl::c_sort instead of std::c_sort.
Best-effort to copy the backend_config over when replacing HloInstructions (But only if the new one is empty)

mellanox-github commented:
Can one of the admins verify this patch?

AnthonyBarbier committed:
Actually there is no need to copy over the backend config at the Hlo level

sanjoy reviewed (dismissed):
lgtm

AnthonyBarbier committed:
Ran buildifier to fix formatting issues in BUILD files

sanjoy reviewed (dismissed):
lgtm

AnthonyBarbier committed:
Fixed xla_builder_test: AddFrontendAttribute -> SetInstructionFrontendAttribute

AnthonyBarbier commented:
FYI: Don't know if it's intentional but it looks like the CI doesn't run or build `//tensorflow/compiler/xla/client:all`

rthadur commented:
@AnthonyBarbier there are some conflicts , can you please resolve

AnthonyBarbier committed:
Merge remote-tracking branch 'master' into frontend_attributes

AnthonyBarbier commented:
> @AnthonyBarbier there are some conflicts , can you please resolve

Sure. Done. There seems to be some issues with LLVM but that's got something to do with this patch.

```
ERROR: Execution platform: @bazel_tools//platforms:host_platform
ERROR: In file included from external/llvm/include/llvm/TableGen/Record.h:27:0,
ERROR:                  from external/llvm/utils/TableGen/WebAssemblyDisassemblerEmitter.cpp:17:
ERROR: external/llvm/include/llvm/Support/TrailingObjects.h: In static member function 'static void llvm::TrailingObjects<BaseTy, TrailingTys>::verifyTrailingObjectsAssertions()':
ERROR: external/llvm/include/llvm/Support/TrailingObjects.h:252:24: error: 'is_final' is not a member of 'std'
ERROR:      static_assert(std::is_final<BaseTy>(), "BaseTy must be final.");
ERROR:                         ^~~~~~~~
ERROR: external/llvm/include/llvm/Support/TrailingObjects.h:252:24: note: suggested alternative: 'is_heap'
ERROR:      static_assert(std::is_final<BaseTy>(), "BaseTy must be final.");
ERROR:                         ^~~~~~~~
ERROR:                         is_heap
ERROR: external/llvm/include/llvm/Support/TrailingObjects.h:252:39: error: expected primary-expression before '>' token
ERROR:      static_assert(std::is_final<BaseTy>(), "BaseTy must be final.");
ERROR:                                        ^
ERROR: external/llvm/include/llvm/Support/TrailingObjects.h:252:41: error: expected primary-expression before ')' token
ERROR:      static_assert(std::is_final<BaseTy>(), "BaseTy must be final.");
ERROR:                                          ^
ERROR: /localdata/anthonyb/workspace/tensorflowview/tensorflow/tensorflow/compiler/xla/service/cpu/BUILD:71:1 C++ compilation of rule '@llvm//:llvm-tblgen' failed (Exit 1): gcc failed: error executing command
ERROR:   (cd /localdata/anthonyb/cache/bazel/_bazel_anthonyb/588fec831672060331cd1471480aae44/execroot/org_tensorflow && \
ERROR:   exec env - \
ERROR:     PATH=/home/anthonyb/work/gdb/bin:/usr/lib/ccache:/home/anthonyb/work/vimrc/install/bin:/home/anthonyb/bin:/home/anthonyb/work/arcanist/bin/:/home/anthonyb/.local/bin:/home/anthonyb/bin:/home/anthonyb/work/gdb/bin:/usr/lib/ccache:/home/anthonyb/work/vimrc/install/bin:/home/anthonyb/bin:/home/anthonyb/work/arcanist/bin/:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/home/anthonyb/.fzf/bin \
ERROR:     PWD=/proc/self/cwd \
ERROR:   /usr/lib/ccache/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections '-std=c++0x' -MD -MF bazel-out/host/bin/external/llvm/_objs/llvm-tblgen/WebAssemblyDisassemblerEmitter.d '-frandom-seed=bazel-out/host/bin/external/llvm/_objs/llvm-tblgen/WebAssemblyDisassemblerEmitter.o' -DLLVM_ENABLE_STATS -D__STDC_LIMIT_MACROS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -DLLVM_BUILD_GLOBAL_ISEL -iquote external/llvm -iquote bazel-out/host/genfiles/external/llvm -iquote bazel-out/host/bin/external/llvm -iquote external/zlib_archive -iquote bazel-out/host/genfiles/external/zlib_archive -iquote bazel-out/host/bin/external/zlib_archive -iquote external/bazel_tools -iquote bazel-out/host/genfiles/external/bazel_tools -iquote bazel-out/host/bin/external/bazel_tools -isystem external/llvm/include -isystem bazel-out/host/genfiles/external/llvm/include -isystem bazel-out/host/bin/external/llvm/include -isystem external/zlib_archive -isystem bazel-out/host/genfiles/external/zlib_archive -isystem bazel-out/host/bin/external/zlib_archive -g0 -w '-march=native' -g0 -fno-canonical-system-headers -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c external/llvm/utils/TableGen/WebAssemblyDisassemblerEmitter.cpp -o bazel-out/host/bin/external/llvm/_objs/llvm-tblgen/WebAssemblyDisassemblerEmitter.o)
```

AnthonyBarbier commented:
Did someone accidentally merged this PR ? 
Looks like it wasn't the latest version though that get merged.
